customer/Customer.cpp

```
@@ -0,0 +1,118 @@
+#include "Customer.h"
+#include <iostream>
+
+Customer::Customer(std::string _name, int _age)
+    : name(_name), age(_age) {}
+
+Customer::~Customer() {}
+
+RegularCustomer::RegularCustomer(std::string _name, int _age)
+    : Customer(_name, _age)
+{
+    if (_age < 18)
+    {
+        throw std::invalid_argument("Regular customer's age must be grater or equal to 18");
+    }
+}
+
+double RegularCustomer::getTotalCost()
+{
+    double total = 0;
+
+    for (auto &pair : shopping_cart)
+    {
+        total += pair.first->getPrice() * pair.second;
+    }
+
+    return total;
+}
+
+void RegularCustomer::buyProduct(Store &s, std::shared_ptr<Product> p, int quantity)
+{
+    try
+    {
+        s.purchase(p, quantity, p->getPrice() * quantity);
+        shopping_cart[p] += quantity;
+    }
+    catch (std::invalid_argument &e)
+    {
+        std::cout << e.what() << '\n';
+    }
```

It doesn't make sense for the customer to eat the exception here. The buy failed, but the function that calls buyProduct never finds out. And the buyProduct function doesn't do anything to try recover from the error.